Импортируем pandas, загружаем датасет "качество воды", в который входит информация о веществах, которые входят в состав воды из различных образцов ( алюминий, аммиак, мышьяк, барий, кадмий, хлорамин, хром, фторид меди, бактерии, нитраты свинца, нитриты, перхлорат ртути, радий, селен, серебро, уран.), а также, безопасна вода или нет

In [2]:
import pandas as pd
per = pd.read_csv("waterQuality1.csv")
#грузим воду
per

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  \
0          1.65    9.08     0.04    2.85    0.007        0.35      0.83   
1          2.32   21.16     0.01    3.31    0.002        5.28      0.68   
2          1.01   14.02     0.04    0.58    0.008        4.24      0.53   
3          1.36   11.33     0.04    2.96    0.001        7.23      0.03   
4          0.92   24.33     0.03    0.20    0.006        2.67      0.69   
...         ...     ...      ...     ...      ...         ...       ...   
7994       0.05    7.78     0.00    1.95    0.040        0.10      0.03   
7995       0.05   24.22     0.02    0.59    0.010        0.45      0.02   
7996       0.09    6.85     0.00    0.61    0.030        0.05      0.05   
7997       0.01      10     0.01    2.00    0.000        2.00      0.00   
7998       0.04    6.85     0.01    0.70    0.030        0.05      0.01   

      copper  flouride  bacteria  ...   lead  nitrates  nitrites  mercury  \
0       0.17      0.05      0.20  ...  0.054     16.08      1.13    0.007   
1       0.66      0.90      0.65  ...  0.100      2.01      1.93    0.003   
2       0.02      0.99      0.05  ...  0.078     14.16      1.11    0.006   
3       1.66      1.08      0.71  ...  0.016      1.41      1.29    0.004   
4       0.57      0.61      0.13  ...  0.117      6.74      1.11    0.003   
...      ...       ...       ...  ...    ...       ...       ...      ...   
7994    0.03      1.37      0.00  ...  0.197     14.29      1.00    0.005   
7995    0.02      1.48      0.00  ...  0.031     10.27      1.00    0.001   
7996    0.02      0.91      0.00  ...  0.182     15.92      1.00    0.000   
7997    0.09      0.00      0.00  ...  0.000      0.00      0.00    0.000   
7998    0.03      1.00      0.00  ...  0.182     15.92      1.00    0.000   

      perchlorate  radium  selenium  silver  uranium  is_safe  
0           37.75    6.78      0.08    0.34     0.02        1  
1           32.26    3.21      0.08    0.27     0.05        1  
2           50.28    7.07      0.07    0.44     0.01        0  
3            9.12    1.72      0.02    0.45     0.05        1  
4           16.90    2.41      0.02    0.06     0.02        1  
...           ...     ...       ...     ...      ...      ...  
7994         3.57    2.13      0.09    0.06     0.03        1  
7995         1.48    1.11      0.09    0.10     0.08        1  
7996         1.35    4.84      0.00    0.04     0.05        1  
7997         0.00    0.00      0.00    0.00     0.00        1  
7998         1.35    4.84      0.00    0.04     0.05        1  

[7999 rows x 21 columns]

 С помощью  shape  мы получаем количество строк и столбцов в датасете. В данном датасете 7999 строк и 21 столбец.

In [6]:
per.shape
#получаем размерность вады

(7999, 21)

Здесь мы выводим названя столбцов

In [3]:
print(per.columns)
#получаем названия вады

Index(['aluminium', 'ammonia', 'arsenic', 'barium', 'cadmium', 'chloramine',
       'chromium', 'copper', 'flouride', 'bacteria', 'viruses', 'lead',
       'nitrates', 'nitrites', 'mercury', 'perchlorate', 'radium', 'selenium',
       'silver', 'uranium', 'is_safe'],
      dtype='object')


Здесь выводим уникальные значения из столбцов 

In [4]:
for column in per.columns:
    print(per[column].unique())
    per
    #понимаем, что вода у нас не красивая

[1.65 2.32 1.01 1.36 0.92 0.94 2.36 3.93 0.6  0.22 3.27 1.35 1.88 4.93
 2.89 0.61 3.47 2.11 4.88 4.12 0.68 1.15 0.27 4.32 3.31 1.82 3.42 4.41
 4.57 1.69 1.87 2.71 1.63 0.01 4.49 3.52 4.35 2.51 2.34 1.71 3.45 3.37
 3.13 0.29 1.51 1.09 2.68 0.3  0.23 2.55 0.86 0.24 3.63 3.26 0.42 0.77
 0.58 3.82 2.78 3.62 3.88 4.64 2.25 0.54 4.62 2.5  3.08 3.03 2.85 1.38
 2.73 3.02 1.96 2.35 2.45 2.67 1.12 3.3  0.43 1.19 0.85 0.5  4.75 1.89
 3.75 3.32 1.75 2.66 2.76 2.63 0.04 0.88 2.22 4.91 4.8  0.47 1.62 0.09
 4.66 2.74 1.79 4.15 3.78 2.91 4.3  3.24 3.53 3.04 2.21 2.47 1.92 1.22
 2.12 4.19 4.21 4.07 2.56 3.55 3.94 0.8  2.59 3.9  3.15 0.18 3.86 1.25
 2.29 0.07 4.18 4.34 2.39 0.9  0.89 2.4  3.69 4.71 3.36 4.46 4.5  0.57
 2.69 2.17 1.32 5.   2.6  0.35 3.83 2.9  1.56 2.23 3.81 2.75 0.81 4.
 4.08 1.55 2.42 2.99 1.43 1.66 3.38 4.42 3.43 3.65 4.86 2.82 1.05 0.73
 1.21 2.15 3.6  1.73 3.35 0.37 0.79 0.99 1.86 4.84 3.09 1.72 3.29 0.11
 4.04 3.71 3.79 3.05 0.84 1.98 0.59 0.36 1.4  1.46 2.02 0.   0.93 0.78
 4.48 0.

Здесь мы сортируем значения в столбце aluminium

In [5]:
per = per.sort_values("aluminium")
per
#сортировка воды

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  \
3386       0.00   28.45     0.53    3.79     0.10        6.34      0.88   
6126       0.00   23.15     0.01    0.92     0.01        0.27      0.02   
6120       0.00   22.07     0.02    0.06     0.01        0.10      0.05   
4255       0.00    7.09     0.03    0.04     0.06        0.18      0.04   
7447       0.00    8.39     0.03    0.51     0.08        0.02      0.01   
...         ...     ...      ...     ...      ...         ...       ...   
1490       5.01   19.29     0.65    3.45     0.03        1.82      0.64   
1831       5.03    7.38     0.66    4.21     0.10        1.31      0.81   
1501       5.03   26.11     0.06    1.09     0.04        6.07      0.73   
1613       5.03    7.59     0.13    0.94     0.04        3.22      0.19   
1167       5.05   19.35     0.28    2.90     0.05        1.02      0.19   

      copper  flouride  bacteria  ...   lead  nitrates  nitrites  mercury  \
3386    1.23      1.48      0.57  ...  0.106      7.17      1.23    0.005   
6126    1.95      1.38      0.46  ...  0.191     18.27      1.03    0.009   
6120    0.24      0.66      0.09  ...  0.170     16.17      1.51    0.006   
4255    0.10      0.45      0.00  ...  0.198      7.10      0.21    0.008   
7447    1.67      0.02      0.10  ...  0.028     14.45      2.26    0.009   
...      ...       ...       ...  ...    ...       ...       ...      ...   
1490    0.36      0.64      0.00  ...  0.005      2.01      1.69    0.001   
1831    1.52      0.94      0.15  ...  0.042      2.08      2.27    0.008   
1501    0.87      0.29      0.00  ...  0.195      8.77      2.11    0.007   
1613    1.00      0.36      0.00  ...  0.001      1.17      1.41    0.008   
1167    0.36      1.39      0.00  ...  0.116      3.25      1.80    0.001   

      perchlorate  radium  selenium  silver  uranium  is_safe  
3386        17.87    7.71      0.09    0.09     0.04        0  
6126         2.27    0.99      0.05    0.05     0.02        0  
6120         1.05    0.97      0.02    0.04     0.02        0  
4255         2.05    0.56      0.00    0.05     0.06        0  
7447         0.38    5.18      0.06    0.04     0.03        0  
...           ...     ...       ...     ...      ...      ...  
1490        16.50    1.45      0.07    0.32     0.05        0  
1831         3.20    1.71      0.05    0.49     0.02        0  
1501         1.42    7.52      0.06    0.26     0.02        0  
1613        47.79    3.80      0.09    0.32     0.04        0  
1167         1.29    7.85      0.06    0.05     0.08        0  

[7999 rows x 21 columns]

Здесь мы удаляем дубликаты значений 

In [6]:
per = per.drop_duplicates()
per
#удаляем дубликаты ^_^

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  \
3386       0.00   28.45     0.53    3.79     0.10        6.34      0.88   
6126       0.00   23.15     0.01    0.92     0.01        0.27      0.02   
6120       0.00   22.07     0.02    0.06     0.01        0.10      0.05   
4255       0.00    7.09     0.03    0.04     0.06        0.18      0.04   
7447       0.00    8.39     0.03    0.51     0.08        0.02      0.01   
...         ...     ...      ...     ...      ...         ...       ...   
1490       5.01   19.29     0.65    3.45     0.03        1.82      0.64   
1831       5.03    7.38     0.66    4.21     0.10        1.31      0.81   
1501       5.03   26.11     0.06    1.09     0.04        6.07      0.73   
1613       5.03    7.59     0.13    0.94     0.04        3.22      0.19   
1167       5.05   19.35     0.28    2.90     0.05        1.02      0.19   

      copper  flouride  bacteria  ...   lead  nitrates  nitrites  mercury  \
3386    1.23      1.48      0.57  ...  0.106      7.17      1.23    0.005   
6126    1.95      1.38      0.46  ...  0.191     18.27      1.03    0.009   
6120    0.24      0.66      0.09  ...  0.170     16.17      1.51    0.006   
4255    0.10      0.45      0.00  ...  0.198      7.10      0.21    0.008   
7447    1.67      0.02      0.10  ...  0.028     14.45      2.26    0.009   
...      ...       ...       ...  ...    ...       ...       ...      ...   
1490    0.36      0.64      0.00  ...  0.005      2.01      1.69    0.001   
1831    1.52      0.94      0.15  ...  0.042      2.08      2.27    0.008   
1501    0.87      0.29      0.00  ...  0.195      8.77      2.11    0.007   
1613    1.00      0.36      0.00  ...  0.001      1.17      1.41    0.008   
1167    0.36      1.39      0.00  ...  0.116      3.25      1.80    0.001   

      perchlorate  radium  selenium  silver  uranium  is_safe  
3386        17.87    7.71      0.09    0.09     0.04        0  
6126         2.27    0.99      0.05    0.05     0.02        0  
6120         1.05    0.97      0.02    0.04     0.02        0  
4255         2.05    0.56      0.00    0.05     0.06        0  
7447         0.38    5.18      0.06    0.04     0.03        0  
...           ...     ...       ...     ...      ...      ...  
1490        16.50    1.45      0.07    0.32     0.05        0  
1831         3.20    1.71      0.05    0.49     0.02        0  
1501         1.42    7.52      0.06    0.26     0.02        0  
1613        47.79    3.80      0.09    0.32     0.04        0  
1167         1.29    7.85      0.06    0.05     0.08        0  

[7999 rows x 21 columns]

Здесь мы заменяем все значения 0.00 в столбце aluminium на значение 1

In [7]:
per["aluminium"] = per["aluminium"].replace(0.00, 1)
per

aluminium ammonia  arsenic  barium  cadmium  chloramine  chromium  \
3386       1.00   28.45     0.53    3.79     0.10        6.34      0.88   
6126       1.00   23.15     0.01    0.92     0.01        0.27      0.02   
6120       1.00   22.07     0.02    0.06     0.01        0.10      0.05   
4255       1.00    7.09     0.03    0.04     0.06        0.18      0.04   
7447       1.00    8.39     0.03    0.51     0.08        0.02      0.01   
...         ...     ...      ...     ...      ...         ...       ...   
1490       5.01   19.29     0.65    3.45     0.03        1.82      0.64   
1831       5.03    7.38     0.66    4.21     0.10        1.31      0.81   
1501       5.03   26.11     0.06    1.09     0.04        6.07      0.73   
1613       5.03    7.59     0.13    0.94     0.04        3.22      0.19   
1167       5.05   19.35     0.28    2.90     0.05        1.02      0.19   

      copper  flouride  bacteria  ...   lead  nitrates  nitrites  mercury  \
3386    1.23      1.48      0.57  ...  0.106      7.17      1.23    0.005   
6126    1.95      1.38      0.46  ...  0.191     18.27      1.03    0.009   
6120    0.24      0.66      0.09  ...  0.170     16.17      1.51    0.006   
4255    0.10      0.45      0.00  ...  0.198      7.10      0.21    0.008   
7447    1.67      0.02      0.10  ...  0.028     14.45      2.26    0.009   
...      ...       ...       ...  ...    ...       ...       ...      ...   
1490    0.36      0.64      0.00  ...  0.005      2.01      1.69    0.001   
1831    1.52      0.94      0.15  ...  0.042      2.08      2.27    0.008   
1501    0.87      0.29      0.00  ...  0.195      8.77      2.11    0.007   
1613    1.00      0.36      0.00  ...  0.001      1.17      1.41    0.008   
1167    0.36      1.39      0.00  ...  0.116      3.25      1.80    0.001   

      perchlorate  radium  selenium  silver  uranium  is_safe  
3386        17.87    7.71      0.09    0.09     0.04        0  
6126         2.27    0.99      0.05    0.05     0.02        0  
6120         1.05    0.97      0.02    0.04     0.02        0  
4255         2.05    0.56      0.00    0.05     0.06        0  
7447         0.38    5.18      0.06    0.04     0.03        0  
...           ...     ...       ...     ...      ...      ...  
1490        16.50    1.45      0.07    0.32     0.05        0  
1831         3.20    1.71      0.05    0.49     0.02        0  
1501         1.42    7.52      0.06    0.26     0.02        0  
1613        47.79    3.80      0.09    0.32     0.04        0  
1167         1.29    7.85      0.06    0.05     0.08        0  

[7999 rows x 21 columns]

Здесь мы удаляем столбец ammonia

In [11]:
per = per.drop("ammonia",axis=1)
per

aluminium  arsenic  barium  cadmium  chloramine  chromium  copper  \
3386       1.00     0.53    3.79     0.10        6.34      0.88    1.23   
6126       1.00     0.01    0.92     0.01        0.27      0.02    1.95   
6120       1.00     0.02    0.06     0.01        0.10      0.05    0.24   
4255       1.00     0.03    0.04     0.06        0.18      0.04    0.10   
7447       1.00     0.03    0.51     0.08        0.02      0.01    1.67   
...         ...      ...     ...      ...         ...       ...     ...   
1490       5.01     0.65    3.45     0.03        1.82      0.64    0.36   
1831       5.03     0.66    4.21     0.10        1.31      0.81    1.52   
1501       5.03     0.06    1.09     0.04        6.07      0.73    0.87   
1613       5.03     0.13    0.94     0.04        3.22      0.19    1.00   
1167       5.05     0.28    2.90     0.05        1.02      0.19    0.36   

      flouride  bacteria  viruses   lead  nitrates  nitrites  mercury  \
3386      1.48      0.57    0.570  0.106      7.17      1.23    0.005   
6126      1.38      0.46    0.003  0.191     18.27      1.03    0.009   
6120      0.66      0.09    0.008  0.170     16.17      1.51    0.006   
4255      0.45      0.00    0.790  0.198      7.10      0.21    0.008   
7447      0.02      0.10    0.007  0.028     14.45      2.26    0.009   
...        ...       ...      ...    ...       ...       ...      ...   
1490      0.64      0.00    0.690  0.005      2.01      1.69    0.001   
1831      0.94      0.15    0.008  0.042      2.08      2.27    0.008   
1501      0.29      0.00    0.580  0.195      8.77      2.11    0.007   
1613      0.36      0.00    0.005  0.001      1.17      1.41    0.008   
1167      1.39      0.00    0.000  0.116      3.25      1.80    0.001   

      perchlorate  radium  selenium  silver  uranium is_safe  
3386        17.87    7.71      0.09    0.09     0.04       0  
6126         2.27    0.99      0.05    0.05     0.02       0  
6120         1.05    0.97      0.02    0.04     0.02       0  
4255         2.05    0.56      0.00    0.05     0.06       0  
7447         0.38    5.18      0.06    0.04     0.03       0  
...           ...     ...       ...     ...      ...     ...  
1490        16.50    1.45      0.07    0.32     0.05       0  
1831         3.20    1.71      0.05    0.49     0.02       0  
1501         1.42    7.52      0.06    0.26     0.02       0  
1613        47.79    3.80      0.09    0.32     0.04       0  
1167         1.29    7.85      0.06    0.05     0.08       0  

[7999 rows x 20 columns]

Получаем общую информацию о датасете, количество непустых значений, типы данных в каждом столбце и объём потребляемой памяти 

In [7]:
per.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7999 entries, 0 to 7998
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   aluminium    7999 non-null   float64
 1   ammonia      7999 non-null   object 
 2   arsenic      7999 non-null   float64
 3   barium       7999 non-null   float64
 4   cadmium      7999 non-null   float64
 5   chloramine   7999 non-null   float64
 6   chromium     7999 non-null   float64
 7   copper       7999 non-null   float64
 8   flouride     7999 non-null   float64
 9   bacteria     7999 non-null   float64
 10  viruses      7999 non-null   float64
 11  lead         7999 non-null   float64
 12  nitrates     7999 non-null   float64
 13  nitrites     7999 non-null   float64
 14  mercury      7999 non-null   float64
 15  perchlorate  7999 non-null   float64
 16  radium       7999 non-null   float64
 17  selenium     7999 non-null   float64
 18  silver       7999 non-null   float64
 19  uraniu

Получаем статическое описание числовых данных датасета 

In [8]:
per.describe()

aluminium      arsenic       barium      cadmium   chloramine  \
count  7999.000000  7999.000000  7999.000000  7999.000000  7999.000000   
mean      0.666158     0.161445     1.567715     0.042806     2.176831   
std       1.265145     0.252590     1.216091     0.036049     2.567027   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.040000     0.030000     0.560000     0.008000     0.100000   
50%       0.070000     0.050000     1.190000     0.040000     0.530000   
75%       0.280000     0.100000     2.480000     0.070000     4.240000   
max       5.050000     1.050000     4.940000     0.130000     8.680000   

          chromium       copper     flouride     bacteria      viruses  \
count  7999.000000  7999.000000  7999.000000  7999.000000  7999.000000   
mean      0.247226     0.805857     0.771565     0.319665     0.328583   
std       0.270640     0.653539     0.435373     0.329485     0.378096   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.050000     0.090000     0.405000     0.000000     0.002000   
50%       0.090000     0.750000     0.770000     0.220000     0.008000   
75%       0.440000     1.390000     1.160000     0.610000     0.700000   
max       0.900000     2.000000     1.500000     1.000000     1.000000   

              lead     nitrates     nitrites      mercury  perchlorate  \
count  7999.000000  7999.000000  7999.000000  7999.000000  7999.000000   
mean      0.099450     9.818822     1.329961     0.005194    16.460299   
std       0.058172     5.541331     0.573219     0.002967    17.687474   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.048000     5.000000     1.000000     0.003000     2.170000   
50%       0.102000     9.930000     1.420000     0.005000     7.740000   
75%       0.151000    14.610000     1.760000     0.008000    29.480000   
max       0.200000    19.830000     2.930000     0.010000    60.010000   

            radium     selenium       silver      uranium  
count  7999.000000  7999.000000  7999.000000  7999.000000  
mean      2.920548     0.049685     0.147781     0.044673  
std       2.323009     0.028770     0.143551     0.026904  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.820000     0.020000     0.040000     0.020000  
50%       2.410000     0.050000     0.080000     0.050000  
75%       4.670000     0.070000     0.240000     0.070000  
max       7.990000     0.100000     0.500000     0.090000

Здесь мы вовидм значения первой строчки у каждого столбца 

In [10]:
row_data = per.loc[1]
row_data

aluminium       2.32
ammonia        21.16
arsenic         0.01
barium          3.31
cadmium        0.002
chloramine      5.28
chromium        0.68
copper          0.66
flouride         0.9
bacteria        0.65
viruses         0.65
lead             0.1
nitrates        2.01
nitrites        1.93
mercury        0.003
perchlorate    32.26
radium          3.21
selenium        0.08
silver          0.27
uranium         0.05
is_safe            1
Name: 1, dtype: object

In [9]:
per.to_csv("newVoda.csv")